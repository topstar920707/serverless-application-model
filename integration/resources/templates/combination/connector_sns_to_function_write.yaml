Resources:
  Topic:
    Type: AWS::SNS::Topic
    Properties:
      Subscription:
        - Endpoint: !GetAtt 'InvokedFunction.Arn'
          Protocol: lambda

  TriggerFunction:
    Type: AWS::Serverless::Function
    Properties:
      Runtime: nodejs14.x
      Handler: index.handler
      Timeout: 10  # in case eb has delay
      InlineCode: |
          const AWS = require('aws-sdk');
          exports.handler = async (event) => {
            var params = {
              Message: 'Test Message',
              TopicArn: process.env.TOPIC_ARN,
            };
            var sns = new AWS.SNS();
            var sqs = new AWS.SQS();
            await sns.publish(params).promise();

            const data = await sqs.receiveMessage({
              QueueUrl: process.env.VERIFICATION_QUEUE_URL,
              WaitTimeSeconds: 5,
            }).promise();
            if (data.Messages.length == 0) {
              throw 'No messages in the queue!';
            }
          };
      Environment:
        Variables:
          TOPIC_ARN: !Ref Topic
          VERIFICATION_QUEUE_URL: !Ref VerificationQueue
          EVENT_SOURCE: !Sub '${AWS::StackName}-test-event'
      Policies:
        - SQSPollerPolicy:
            QueueName: !GetAtt VerificationQueue.QueueName

  InvokedFunction:
    Type: AWS::Serverless::Function
    Properties:
      Runtime: nodejs14.x
      Handler: index.handler
      InlineCode: |
        const AWS = require('aws-sdk');
        exports.handler = async (event) => {
          const sqs = new AWS.SQS();
          await sqs.sendMessage({
            QueueUrl: process.env.VERIFICATION_QUEUE_URL,
            MessageBody: "test"
          }).promise();
        };
      Environment:
        Variables:
          VERIFICATION_QUEUE_URL: !Ref VerificationQueue
      Policies:
        - SQSSendMessagePolicy:
            QueueName: !GetAtt VerificationQueue.QueueName
  
  VerificationQueue:
    Type: AWS::SQS::Queue

  Connector:
    Type: AWS::Serverless::Connector
    Properties:
      Source: 
        Id: Topic
      Destination: 
        Id: InvokedFunction
      Permissions:
        - Write

  ConnectorNotTested:
    Type: AWS::Serverless::Connector
    Properties:
      Source:
        Id: TriggerFunction
      Destination: 
        Id: Topic
      Permissions:
        - Write

