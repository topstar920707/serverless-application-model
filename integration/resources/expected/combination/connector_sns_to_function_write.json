[
  {
    "LogicalResourceId": "TriggerFunctionRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "TriggerFunction",
    "ResourceType": "AWS::Lambda::Function"
  },
  {
    "LogicalResourceId": "InvokedFunctionRole",
    "ResourceType": "AWS::IAM::Role"
  },
  {
    "LogicalResourceId": "InvokedFunction",
    "ResourceType": "AWS::Lambda::Function"
  },
  {
    "LogicalResourceId": "Topic",
    "ResourceType": "AWS::SNS::Topic"
  },
  {
    "LogicalResourceId": "VerificationQueue",
    "ResourceType": "AWS::SQS::Queue"
  },
  {
    "LogicalResourceId": "ConnectorWriteLambdaPermission",
    "ResourceType": "AWS::Lambda::Permission"
  },
  {
    "LogicalResourceId": "ConnectorNotTestedPolicy",
    "ResourceType": "AWS::IAM::ManagedPolicy"
  }
]